### ISSUE TITLE
Spoa-mirror crashes with segfault

### ISSUE THREAD

**Issue by mazhora** [haproxytech/spoa-mirror#31]
When we try to mirror traffic spoa-mirror crashes with segfault when receiving traffic. This problem exists on ubuntu 22.
Everything works fine with the same settings on ubuntu 20. I can't figure out what the problem is.
Can you help me to fix it ?

Some debug information

```
# dmesg
...
[3026605.955875] sm/wrk: 1[1866406]: segfault at 40 ip 00007f33ad6e30ee sp 00007f33ac5476d0 error 4 in libcurl.so.4.7.0[7f33ad6a9000+74000]
[3026605.955885] Code: 00 48 ba ff ff ff ff ff ff ff 7f 0f 85 7a ef ff ff e9 32 ff ff ff 4d 8b 87 58 0d 00 00 4c 89 e1 48 b8 f6 53 e3 a5 9b c4 20 00 <49> 2b 48 40 41 8b 50 48 48 39 c1 0f 8f 0d ff ff ff 48 b8 0a ac 1c
```

```
# journalctl -ex -u haproxy
...
Sep 14 09:46:38 staging-front-1-vin haproxy[1866392]: [ALERT]    (1866392) : Current program 'mirror' (1866404) exited with code 139 (Segmentation fault)
Sep 14 09:46:38 staging-front-1-vin haproxy[1866392]: [ALERT]    (1866392) : exit-on-failure: killing every processes with SIGTERM
...
```

```
# spoa-mirror -V
spoa-mirror v1.2.15 [build 2495] by Miroslav Zagorac <mzagorac@haproxy.com>, Sep 13 2022
```

```
# haproxy -v
HAProxy version 2.4.14-1ubuntu1 2022/02/28 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.14.html
Running on: Linux 5.15.0-43-generic #46-Ubuntu SMP Tue Jul 12 10:30:17 UTC 2022 x86_64
```

```
# lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.1 LTS
Release:	22.04
Codename:	jammy
```

```
/etc/haproxy/haproxy.conf
...
program mirror
 command spoa-mirror --runtime 0 --mirror-url http://10.99.2.50:8000
...
frontend fe_https_in
 filter spoe engine mirror config /etc/haproxy/mirror.conf
...
backend mirroragents
  mode tcp
  balance roundrobin
  timeout connect 5s
  timeout server 5s
  server agent1 localhost:12345
...
```

```
/etc/haproxy/mirror.conf
[mirror]
spoe-agent mirror
  log global
  messages mirror
  use-backend mirroragents
  timeout hello 500ms
  timeout idle 5s
  timeout processing 5s

spoe-message mirror
  args arg_method=method arg_path=url arg_ver=req.ver arg_hdrs=req.hdrs_bin arg_body=req.body
  event on-frontend-http-request
```

**Comment by baylanger** [comment]
If libcurl.so.4.7.0 is on U22 , what release is on U20?  You could try to downgrade if needed.  It's possible 4.7 introduced changes in their functions that requires adjustment in haproxy.

strace of the process crashing might help give a clue.  Otherwise a compiled version with debug info could produce a valuable core dump.

**Comment by zaga00** [comment]
Hello @mazhora,

I tried spoa-mirror on the ubuntu 22 system and I confirm that it behaves the same as yours (the program crashes).  Unfortunately, I can't spend much time looking for bugs right now, but I hope it will be resolved in a day or two.

**Comment by mazhora** [comment]
Hi @zaga00 ,
I tried the new patched version (v1.2.16) and it works fine. Thank you very much for the quick fix.

**Comment by zaga00** [comment]
Thank you for reporting the bug.
